[CSS Math Functions] Correctly serialize sin/cos/tan functions

nt1m opened the pull request:
#### 15a8c32776708f07ae1c3b9539640cb00ae1e6a8
<pre>
[CSS Math Functions] Correctly serialize sin/cos/tan functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=259011">https://bugs.webkit.org/show_bug.cgi?id=259011</a>
rdar://111947212

Reviewed by Darin Adler.

There are 2 main changes:

- We now always try to resolve top level sin/cos/tan functions, cos(0) will serialize as calc(1) instead of cos(0) for instance. This is consistent with how calc(cos(0)) is serialized.

From the spec <a href="https://drafts.csswg.org/css-values-4/#calc-simplification">https://drafts.csswg.org/css-values-4/#calc-simplification</a>:

&gt; If root is an operator node that’s not one of the calc-operator nodes, and all of its calculation children are numeric values with enough information to compute the operation root represents, return the result of running root’s operation using its children, expressed in the result’s canonical unit.

- tan(90deg) &amp; tan(-90deg) now return respectively calc(infinity) &amp; calc(-infinity)

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/sin-cos-tan-serialize-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/sin-cos-tan-serialize.html:

Update these tests from upstream WPT.

* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::evaluateOperator):
* Source/WebCore/css/calc/CSSCalcOperationNode.h:
* Source/WebCore/platform/calc/CalcExpressionOperation.cpp:
(WebCore::CalcExpressionOperation::evaluate const):

Canonical link: <a href="https://commits.webkit.org/265881@main">https://commits.webkit.org/265881@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ca55ae3b118c372985062c3d8a126ddfca4afa7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12138 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12485 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12787 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13884 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11713 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12169 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14900 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12499 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/14400 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12302 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/13109 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/10278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14300 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10395 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/11019 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18136 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11476 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/11178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/14354 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11668 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9618 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10884 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/10894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/2977 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15210 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1355 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11521 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
```diff
@@ -1369,6 +1370,14 @@ double CSSCalcOperationNode::evaluateOperator(CalcOperator op, const Vector<doub
+        double x = std::fmod(children[0], piDouble * 2);
+        // std::fmod can return negative values.
+        x = x < 0 ? piDouble * 2 + x : x;
+        ASSERT(x >= 0 && x <= piDouble * 2 || std::isnan(x));
```
Could write this:
```
    ASSERT(!(x < 0));
    ASSERT(!(x > piDouble * 2));
```
I like separate assertions so you can tell which one fired instead of `&&`. I also prefer handling `NaN` without an explicit check by taking advantage of how its behavior is defined. But not sure you prefer either of these.

nt1m replied:
Sure I can write this if it handles NaN correctly.

darinadler commented on Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
```diff
@@ -1369,6 +1370,14 @@ double CSSCalcOperationNode::evaluateOperator(CalcOperator op, const Vector<doub
             return std::numeric_limits<double>::quiet_NaN();
+        double x = std::fmod(children[0], piDouble * 2);
+        // std::fmod can return negative values.
+        x = x < 0 ? piDouble * 2 + x : x;
```
I’m a bit surprised we don’t call `std::tan(x)` below.

nt1m replied:
That works too. I got this code from Blink, but it can be cleaned up a bit.

darinadler commented on Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
```diff
@@ -1369,6 +1370,14 @@ double CSSCalcOperationNode::evaluateOperator(CalcOperator op, const Vector<doub
+        // std::fmod can return negative values.
+        x = x < 0 ? piDouble * 2 + x : x;
+        ASSERT(x >= 0 && x <= piDouble * 2 || std::isnan(x));
+        if (x == piOverTwoDouble)
```
Another way to do this is to check for the additional negative value instead of doing addition to normalize to a positive value.

nt1m replied:
I think it's slightly easier to reason over the smallest range of angles.

darinadler commented on Source/WebCore/platform/calc/CalcExpressionOperation.cpp:
```diff
@@ -132,6 +133,14 @@ float CalcExpressionOperation::evaluate(float maxValue) const
     case CalcOperator::Tan: {
         if (m_children.size() != 1)
             return std::numeric_limits<double>::quiet_NaN();
+        double x = std::fmod(m_children[0]->evaluate(maxValue), piDouble * 2);
```
We should share this code, not write it out twice.

nt1m replied:
This whole file is duplicated unfortunately, we've had a bug where the `hypot()` function was wrong in only one of the two files, so this is something I'd like to address. I have some ideas how to refactor to prevent duplication altogether, but I'd like to do it separately to catch potential regressions.

darinadler commented on Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
```diff
@@ -1369,6 +1370,14 @@ double CSSCalcOperationNode::evaluateOperator(CalcOperator op, const Vector<doub
     case CalcOperator::Tan: {
         if (children.size() != 1)
             return std::numeric_limits<double>::quiet_NaN();
+        double x = std::fmod(children[0], piDouble * 2);
```
I’m not convinced that these special cases are sufficient. Perhaps we need even more. Does `tan(180deg)` correctly return `0`? I don’t see test coverage for that and I am concerned that it does not work correctly for similar reasons to these infinity cases. Separately, perhaps the special cases should be based on the return value of `std::tan` rather than on the arguments.

nt1m replied:
I added testcases at: https://github.com/web-platform-tests/wpt/pull/40934

nt1m committed:
[CSS Math Functions] Correctly serialize sin/cos/tan functions
https://bugs.webkit.org/show_bug.cgi?id=259011
rdar://111947212

Reviewed by Darin Adler.

There are 2 main changes:

- We now always try to resolve top level sin/cos/tan functions, cos(0) will serialize as calc(1) instead of cos(0) for instance. This is consistent with how calc(cos(0)) is serialized.

From the spec https://drafts.csswg.org/css-values-4/#calc-simplification:

> If root is an operator node that’s not one of the calc-operator nodes, and all of its calculation children are numeric values with enough information to compute the operation root represents, return the result of running root’s operation using its children, expressed in the result’s canonical unit.

- tan(90deg) & tan(-90deg) now return respectively calc(infinity) & calc(-infinity)

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/sin-cos-tan-serialize-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/sin-cos-tan-serialize.html:

Update these tests from upstream WPT.

* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::evaluateOperator):
* Source/WebCore/css/calc/CSSCalcOperationNode.h:
* Source/WebCore/platform/calc/CalcExpressionOperation.cpp:
(WebCore::CalcExpressionOperation::evaluate const):

Canonical link: https://commits.webkit.org/265881@main
